pkg/yurtappmanager/controller/nodepool/nodepool_controller.go

```diff
@@ -0,0 +1,600 @@
+limitations under the License.
+*/
+
+package nodepool
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"reflect"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	nodeutil "k8s.io/kubernetes/pkg/controller/util/node"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	appsv1alpha1 "github.com/alibaba/openyurt/pkg/yurtappmanager/apis/apps/v1alpha1"
+	"github.com/alibaba/openyurt/pkg/yurtappmanager/constant"
+	"github.com/alibaba/openyurt/pkg/yurtappmanager/util/gate"
+)
+
+const controllerName = "nodepool-controller"
+
+var concurrentReconciles = 3
+
+// NodePoolReconciler reconciles a NodePool object
+type NodePoolReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	recorder          record.EventRecorder
+	createDefaultPool bool
+}
+
+type NodePoolRelatedAttributes struct {
+	Labels      map[string]string `json:"labels,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+	Taints      []corev1.Taint    `json:"taints,omitempty"`
+}
+
+// Add creates a new NodePool Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, ctx context.Context) error {
+	if !gate.ResourceEnabled(&appsv1alpha1.NodePool{}) {
+		return nil
+	}
+	inf := ctx.Value(constant.ContextKeyCreateDefaultPool)
+	cdp, ok := inf.(bool)
+	if !ok {
+		return errors.New("fail to assert interface to bool for command line option createDefaultPool")
+	}
+	return add(mgr, newReconciler(mgr, cdp))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, createDefaultPool bool) reconcile.Reconciler {
+	return &NodePoolReconciler{
+		Client:            mgr.GetClient(),
+		Scheme:            mgr.GetScheme(),
+		recorder:          mgr.GetEventRecorderFor(controllerName),
+		createDefaultPool: createDefaultPool,
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName,
+		mgr, controller.Options{
+			Reconciler:              r,
+			MaxConcurrentReconciles: concurrentReconciles})
+	if err != nil {
+		return err
+	}
+
+	npr, ok := r.(*NodePoolReconciler)
+	if !ok {
+		return errors.New("fail to assert interface to NodePoolReconciler")
+	}
+
+	if npr.createDefaultPool {
```

Comment:
Should you move this if block to the end of this method (after the watcher has been installed)?

Reply:
the if block has been moved to the end of the method.